db.go

```diff
@@ -485,110 +477,191 @@ func (db *DB) reload() (err error) {
-				continue
-			}
-			corrupted[ulid] = err
-			continue
+	for ulid, err := range corrupted {
+		if _, ok := deletable[ulid]; !ok {
+			return errors.Wrap(err, "unexpected corrupted block")
 		}
-		if db.beyondRetention(meta) {
-			deleteable[meta.ULID] = struct{}{}
+	}
+
+	// All deletable blocks should not be loaded.
+	var (
+		bb         []*Block
+		blocksSize int64
+	)
+	for _, block := range loadable {
+		if _, ok := deletable[block.Meta().ULID]; ok {
+			deletable[block.Meta().ULID] = block
 			continue
 		}
-		for _, b := range meta.Compaction.Parents {
-			deleteable[b.ULID] = struct{}{}
-		}
+		bb = append(bb, block)
+		blocksSize += block.Size()
+
 	}
-	// Blocks we failed to open should all be those we are want to delete anyway.
-	for c, err := range corrupted {
-		if _, ok := deleteable[c]; !ok {
-			return errors.Wrapf(err, "unexpected corrupted block %s", c)
+	loadable = bb
+	db.metrics.blocksBytes.Set(float64(blocksSize))
+
+	sort.Slice(loadable, func(i, j int) bool {
+		return loadable[i].Meta().MaxTime < loadable[j].Meta().MaxTime
+	})
+	if err := validateBlockSequence(loadable); err != nil {
+		return errors.Wrap(err, "invalid block sequence")
+	}
+
+	// Swap new blocks first for subsequently created readers to be seen.
+	db.mtx.Lock()
+	oldBlocks := db.blocks
+	db.blocks = loadable
+	db.mtx.Unlock()
+
+	for _, b := range oldBlocks {
+		if _, ok := deletable[b.Meta().ULID]; ok {
+			deletable[b.Meta().ULID] = b
 		}
 	}
-	// Load new blocks into memory.
+
+	if err := db.deleteBlocks(deletable); err != nil {
+		return err
+	}
+
+	// Garbage collect data in the head if the most recent persisted block
+	// covers data of its current time range.
+	if len(loadable) == 0 {
+		return nil
+	}
+
+	maxt := loadable[len(loadable)-1].Meta().MaxTime
+
+	return errors.Wrap(db.head.Truncate(maxt), "head truncate failed")
+}
+
+func (db *DB) openBlocks() (blocks []*Block, corrupted map[ulid.ULID]error, err error) {
+	corrupted = make(map[ulid.ULID]error)
+
+	dirs, err := blockDirs(db.dir)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "find blocks")
+	}
+
 	for _, dir := range dirs {
-		meta, err := readMetaFile(dir)
+		ulid, err := ulid.Parse(filepath.Base(dir))
 		if err != nil {
-			return errors.Wrapf(err, "read meta information %s", dir)
-		}
-		// Don't load blocks that are scheduled for deletion.
-		if _, ok := deleteable[meta.ULID]; ok {
+			level.Error(db.logger).Log("msg", "not a block dir", "dir", dir)
 			continue
 		}
+
 		// See if we already have the block in memory or open it otherwise.
-		b, ok := db.getBlock(meta.ULID)
+		block, ok := db.getBlock(ulid)
 		if !ok {
-			b, err = OpenBlock(dir, db.chunkPool)
+			block, err = OpenBlock(db.logger, dir, db.chunkPool)
 			if err != nil {
-				return errors.Wrapf(err, "open block %s", dir)
+				corrupted[ulid] = err
 			}
+
 		}
-		blocks = append(blocks, b)
-		opened[meta.ULID] = struct{}{}
+		blocks = append(blocks, block)
```

Comment:
We append a nil block on encountering a corruption, not sure that is wanted.

Reply:
yeah could have an extra check here

```
if block!=nil{
    blocks = append(blocks, block)
}
```

although in `reload()` it will exit if it encounters a corrupted block that is not scheduled for deletion.

```
for ulid, err := range corrupted {
		if _, ok := deletable[ulid]; !ok {
			return errors.Wrap(err, "unexpected corrupted block")
		}
	}
```

Reply:
Yes, but if you see, the blocks passed to `deletableBlocks` will have nil blocks and this will panic: https://github.com/prometheus/tsdb/pull/343/files#diff-7a117cbfd5485ed35efcbb2995c37aa0R585

A `continue` right after https://github.com/prometheus/tsdb/pull/343/files#diff-7a117cbfd5485ed35efcbb2995c37aa0R563 will fix it